system/lib/wasm_worker/library_wasm_worker.c

```diff
@@ -37,20 +40,24 @@ emscripten_wasm_worker_t emscripten_create_wasm_worker(void *stackLowestAddress,
 	assert(emscripten_stack_get_base() <= (uintptr_t)stackLowestAddress || emscripten_stack_get_end() >= (uintptr_t)stackLowestAddress + stackSize
 		&& "When creating a Wasm Worker, its stack should be located either in global data or on the heap, not on the calling thread's own stack!");
 
-	// The Worker's TLS area will be spliced off from the stack region.
-    // We expect TLS area to need to be at most 16 bytes aligned
+	// We expect TLS area to need to be at most 16 bytes aligned
 	assert(__builtin_wasm_tls_align() == 0 || 16 % __builtin_wasm_tls_align() == 0);
 
-#ifndef NDEBUG
-	uint32_t tlsSize = (__builtin_wasm_tls_size() + 15) & -16;
-	assert(stackSize > tlsSize);
-#endif
+	// The Worker's TLS area will be spliced off from the stack region, so the
+	// stack needs to be at least as large as the TLS region.
+	uint32_t tlsSize = ROUND_UP(__builtin_wasm_tls_size(), 16);
+	if (stackSize <= tlsSize) {
+		return 0;
+	}
 
 	return _emscripten_create_wasm_worker(stackLowestAddress, stackSize);
 }
 
 emscripten_wasm_worker_t emscripten_malloc_wasm_worker(uint32_t stackSize)
 {
+	// Add the TLS size to the provided stackSize so that the allocation
+	// will always be large enough to hold the worker TLS data.
+	stackSize += ROUND_UP(__builtin_wasm_tls_size(), 16);
```

Comment:
How does this change pass the wasm worker code size test? https://github.com/emscripten-core/emscripten/blob/main/test/wasm_worker/hello_wasm_worker.c#L16

This change has the awkward result of making the `stackSize` parameter in  `emscripten_malloc_wasm_worker()` and `emscripten_create_wasm_worker()` mean two different things.

I think the consistent across the two would be to rename `uint32_t stackSize` to `uint32_t stackPlusTlsSize`, and then in documentation and examples mention users to call `__builtin_wasm_tls_size()` to obtain the TLS size?

Reply:
Yes, I was thinking something like that would be good idea.

Reply:
Done